Add C++ Bindings

JasonAtClockwork opened the pull request:
# Description of Changes

This adds C++ server bindings (/crate/bindings-cpp) to allow writing C++ 20 modules.

- Emscripten WASM build system integration with CMake
- Macro-based code generation (SPACETIMEDB_TABLE, SPACETIMEDB_REDUCER, etc)
- All SpacetimeDB types supported (primitives, Timestamp, Identity, Uuid, etc)
- Product types via SPACETIMEDB_STRUCT
- Sum types via SPACETIMEDB_ENUM
- Constraints marked with FIELD* macros

# API and ABI breaking changes

None

# Expected complexity level and risk

2 - Doesn't heavily impact any other areas but is complex macro C++ structure to support a similar developer experience, did have a small impact on init command

# Testing

- [x] modules/module-test-cpp - heavily tested every reducer
- [x] modules/benchmarks-cpp - tested through the standalone (~6x faster than C#, ~6x slower than Rust)
- [x] modules/sdk-test-cpp
- [x] modules/sdk-test-procedure-cpp
- [x] modules/sdk-test-view-cpp  
- [x] Wrote several test modules myself
- [x] Quickstart smoketest [Currently in progress]
- [ ] Write Blackholio C++ server module

JasonAtClockwork committed:
Initial commit for bindings-cpp, module-test-cpp, sdk-test-cpp, and benchmarks-cpp

JasonAtClockwork committed:
Updated to latest spacetime init functionality (server only)

JasonAtClockwork committed:
Formatting issue fixed

JasonAtClockwork committed:
Add Cpp to UnrealCpp map for spacetime dev

JasonAtClockwork committed:
Cleared up debug logging, added default field macro, and updated misc_exports to carry defaults

JasonAtClockwork committed:
Fix ahead of views for reducers to properly check for ReducerContext + args

JasonAtClockwork committed:
Updated error handling to an Ok()/Err() style to match closer to Rust as exceptions are not available for C++ yet

JasonAtClockwork committed:
Adding views and the sdk-test-view-cpp

JasonAtClockwork committed:
Start of procedures + procedures with transactions as well as a repair to how Views worked with errors + added a version of IterBuf for C++ to start to reduce allocations

JasonAtClockwork committed:
Added http for C++ procedures

JasonAtClockwork committed:
Added sender_auth and jwt_claims

JasonAtClockwork committed:
Added uuid and result bsatn types

JasonAtClockwork committed:
Cleanup and documentation

JasonAtClockwork committed:
Cleanup and more documentation work

JasonAtClockwork committed:
Fix up for init and build subcommands

JasonAtClockwork committed:
Fixed special reducers to not have a hidden declaration of ReducerContext

JasonAtClockwork committed:
Several repairs to around tests and documentation updates

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

JasonAtClockwork committed:
Temporarily remove hash check for template C++ project till we have a real hosted setup

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

JasonAtClockwork committed:
Hopeful fix for CI + cleanup of templates

JasonAtClockwork committed:
Fixes to benchmarks and tests

clockwork-labs-bot committed:
Update LLM benchmark results

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

# Conflicts:
#	docs/llms/docs-benchmark-analysis.md
#	docs/llms/docs-benchmark-comment.md
#	docs/llms/docs-benchmark-details.json
#	docs/llms/docs-benchmark-summary.json

JasonAtClockwork committed:
Added missing documentation for Ok/Err on reducers and removed requirement for devs to add procedures header manually to match the rest of the architecture (single header)

JasonAtClockwork committed:
Fix for Outcome<> to handle <std:string, std:string> by using a new type to hold the error string as std::variant doesn't allow duplicate types

JasonAtClockwork committed:
Updated sdks/rust/test for C++ and expanded view to work with more than just Rust; as well as fixed bugs with the framework from running tests

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

# Conflicts:
#	sdks/rust/tests/test.rs

JasonAtClockwork committed:
Missing template details

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

# Conflicts:
#	templates/templates-list.json

JasonAtClockwork committed:
Fix the basic-cpp template

JasonAtClockwork committed:
Another template fix for the init subcommand

JasonAtClockwork committed:
Small fixes to QUICKSTART.md

JasonAtClockwork committed:
AI reverted repairs on the documentation, all fixed up again!

JasonAtClockwork committed:
Update crates/bindings-cpp/QUICKSTART.md

Thanks!

Co-authored-by: Ryan <r.ekhoff@clockworklabs.io>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Clean up from Ryan's comments

JasonAtClockwork committed:
Rework HttpClient functions to snake_case matching the other C++ apis

JasonAtClockwork committed:
Fix for init with C++

jdetter reviewed (commented):
Some comments/questions to get started here. There is a lot to review here so I'll do another pass later today.

The main thing I would bring up is that we are referring to this as the "C++ SDK" which I think we should rebrand to the "C++ bindings". My reason is that the other bindings libraries all refer to themselves as bindings, not an SDK. We typically use the nomenclature "SDK" to mean the client SDKs, not the module side bindings.

I fear that if we start referring to this as our C++ SDK this will both confuse users and confuse the LLM training especially if we add a C++ client SDK in the future. If you want feedback from the team on this it might be a good topic to bring up during standup.

Let me know when you're ready for another pass, I'm really excited that this seems like it's working so far and I'm excited for the future with C++26! 🙂

JasonAtClockwork committed:
Update crates/bindings-cpp/ARCHITECTURE.md

Co-authored-by: John Detter <4099508+jdetter@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Update crates/bindings-cpp/ARCHITECTURE.md

Co-authored-by: John Detter <4099508+jdetter@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Initial repairs from John's comments

JasonAtClockwork committed:
Merge branch 'jlarabie/bindings-cpp' of https://github.com/clockworklabs/SpacetimeDB into jlarabie/bindings-cpp

JasonAtClockwork committed:
Rework namespace SpacetimeDb to SpacetimeDB

JasonAtClockwork committed:
More fixes from comments - also fixing more with the init/dev commands

JasonAtClockwork committed:
Added Rust client for basic-cpp template

JasonAtClockwork committed:
Small change to attempt to make accessors more clear

JasonAtClockwork committed:
Added multi-column index accessor

JasonAtClockwork committed:
Fix for `spacetime publish` and C++ detecting the latest binary

rekhoff commented on crates/bindings-cpp/QUICKSTART.md:
```diff
@@ -0,0 +1,293 @@
+
+SpacetimeDB C++ uses a unique accessor pattern:
+- `ctx.db[table_name]` - Access table for iteration and basic operations
+- `ctx.db[table_field]` - Access indexed fields for optimized operations (generated symbol named `tableName_fieldName`)
```
If the goal is to simply clarify how it currently works, my personal preference towards `camelCase`:
```
- `ctx.db[tableName]` - Access table for iteration and basic operations, eg. ctx.db[user]
- `ctx.db[tableName_fieldName]` - Access indexed fields for optimized operations, eg. ctx.db[user_id]
```
This is because it allows it to be clear that we are specifically referring to the name of a given table/field, while also not adding any additional special characters, like underscore, that is present in `snake_case`. The extra underscore get confusing when those are also used as the table to field delimiters.

JasonAtClockwork replied:
Thanks for the feedback I like that and it's in!

JasonAtClockwork committed:
Small lint issue and added view tests back in for C++ through /sdks/rust/tests

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

rekhoff commented on crates/bindings-cpp/ARCHITECTURE.md:
```diff
@@ -0,0 +1,644 @@
+### 2. Outcome<T> - Rust-Like Error Handling
+The SDK provides `Outcome<T>`, a type-safe error handling mechanism matching Rust's `Result<T, E>` pattern:
+- **Outcome<void>** (type alias `ReducerResult`): Used by reducers, can return success (`Ok()`) or error (`Err(message)`)
+- **Outcome<T>**: Used by procedures and other operations, returns either a value or an error message
```
Found another one in the same doc:
Lines 51-52:
```cpp
// For procedures - return a value on success, error message on failure
template<typename T> class Outcome<T>;
```
Looks like this still references a return of `Outcome<T>`

JasonAtClockwork replied:
Not sure if it's like pulling weeds, or playing whack-a-mole here! Thanks I've cleaned up this reference!

JasonAtClockwork committed:
Small documentation tweaks from Ryan's comments

rekhoff reviewed (approved):
Overall LGTM. I very much wish we could use C++26, but given our options, I think the current implementation makes sense. I've reviewed the documentation, skimmed the code, and done some simple local testing. I'd love more eyes on this from core team members, but am happy to sign off approval.

JasonAtClockwork committed:
Updated the index to use datastore_index_scan_point_bsatn and for filter() to return a clean iterator for all use cases
Updated delete_all to work with datastore_delete_by_index_scan_point_bsatn and function correctly

jdetter reviewed (approved):
I've tested this at least with the basic quickstart doc and everything seems to be working. I also looked back and did another pass on the comments that I put down - let me know if you need me to take another look at anything.

I have not done a nit-pick review on the code here since there is quite a lot but I have skimmed it and haven't found anything too concerning. I think it would be better to do a UX review when we go through the follow-up docs PRs anyway. Please request my review when those are ready as well.

Thanks Jason, this is great 👍

JasonAtClockwork committed:
Fix for the `insert_enum_with_payload` test

gefjon reviewed (approved):
I've only looked at the procedure-related stuff in here, but it looks great - much better than I expected, to be honest. Thanks for all your hard work on this!

joshua-spacetime reviewed (approved):
The view code looks correct. `sender` is a method in the 2.0 rust module api. But as I understand it, this is not meant to be 2.0 compliant. Also this does not include the query builder which will be added separately.

JasonAtClockwork committed:
Removal of testing reducers

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

JasonAtClockwork committed:
Cleanup of the bindings-cpp/CMakeLists.txt

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

JasonAtClockwork committed:
Add versioning to the bindings-cpp

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

JasonAtClockwork committed:
Update the version check + upgrade to include C++

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

jdetter reviewed (approved):
I reviewed the changes to `tools/upgrade-version` and the new versioning change, everything looks good here - just a reminder to make sure everything is tested before merging 🙂 

Thanks Jason!

JasonAtClockwork committed:
CI failing due to permissions for cmake

JasonAtClockwork committed:
Another CI fix

JasonAtClockwork committed:
Missed another spot to fix Emscripten in the CI

JasonAtClockwork committed:
Update for the new github hosted spacetimedb-bindings-cpp

JasonAtClockwork committed:
Merge remote-tracking branch 'origin/master' into jlarabie/bindings-cpp

JasonAtClockwork committed:
Compile + build without any debug info for tests

JasonAtClockwork committed:
Merge branch 'master' into jlarabie/bindings-cpp

cloutiertyler reviewed (approved):
Approving my code owned files. Left one comment.

JasonAtClockwork committed:
Fix the LICENSE to be a symlink

JasonAtClockwork committed:
Merge branch 'jlarabie/bindings-cpp' of https://github.com/clockworklabs/SpacetimeDB into jlarabie/bindings-cpp
